File: src/main/java/org/mapdb/ser/ArrayListSerializer.java
package org.mapdb.ser;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.mapdb.io.DataInput2;
import org.mapdb.io.DataOutput2;

import java.util.ArrayList;

public class ArrayListSerializer<E> implements Serializer<ArrayList<E>> {

    protected final Serializer<E> ser;

    public ArrayListSerializer(Serializer<E> ser) {
        this.ser = ser;
    }

    @Override
    public void serialize(@NotNull DataOutput2 out, @NotNull ArrayList<E> list) {
        out.writePackedInt(list.size());
        for(E e:list){
            ser.serialize(out,e);
        }
    }

    @Override
    public ArrayList<E> deserialize(@NotNull DataInput2 input) {
        final int size = input.readPackedInt();
        ArrayList<E> list = new ArrayList<>(size);
        for(int i=0;i<size;i++){
            E e = ser.deserialize(input);
            list.add(e);
        }
        return list;
    }

    @Override
    public @Nullable Class serializedType() {
        return ArrayListSerializer.class;
    }
}

File: src/main/java/org/mapdb/ser/ArrayListSerializer.kt
package org.mapdb.ser

import org.mapdb.io.DataInput2
import org.mapdb.io.DataOutput2

class ArrayListSerializer<K>(val ser: Serializer<K>) : Serializer<ArrayList<K>> {

    override fun serialize(out: DataOutput2, k: ArrayList<K>) {
        out.writePackedInt(k.size)
        for(e in k){
            ser.serialize(out, e)
        }
    }

    override fun deserialize(input: DataInput2): ArrayList<K> {
        val size = input.readPackedInt()
        val list = ArrayList<K>(size)
        for(i in 0 until size){
            val e = ser.deserialize(input)
            list.add(e)
        }
        return list
    }

    override fun serializedType(): Class<*> = ArrayList::class.java

}
